components/dispute-game-contracts/README.md

```diff
@@ -0,0 +1,276 @@
+# Bisection game design
+
+# Context
+
+We assume that the sequencer committed to some transition in the roll-up's state $T = (R, transactions) \to R'$, where R' is the state root (aka `chainState.root`) after applying `transactions` to R, in order.
+
+Suppose a challenger wants to challenge this state. Consider a `MachineState` type something like the following:
+
+```tsx
+type MachineState {
+  root: Bytes32; // hash of (chainState.root, memory.root, instructions)
+  chainState: ChainState;
+  memory: Memory;
+  instructions: Instruction;
+  // etc
+}
+```
+
+(MachineState is assumed to be a Merkle tree, thus the `root` property.)
+
+We force the sequencer to post `S_0.root`, where `S_0 = {chainState: R, memory: emptyMemory, instructions: dissassemble(transactions)}` . The challenger contract must validate that `R` is equal to `R` posted for normal operation rollup. In addition, the challenger contract must validate that `dissassemble(transaction)` is correct. These validations are currently out of scope. We also force the sequencer to post `S'.root`, where `S' = {chainState: R', memory: emptyMemory, instructions: emptyInstructions}` . We might force the sequencer to post additional intermediate states.
+
+We assume there is a function `progress: (state: MachineState) => MachineState` which takes in a machine state `s` and returns the next state. It works by popping the latest instruction, and modifying itself according to the instruction to produce the next state. When `instructions` is empty, `progress(state)` should indicate that it's reached a final state.
+
+The following description is restricted in scope to _correctness_ requirements, and ignores _liveness_ requirements, **assuming that the sequencer & challenger each act in a timely manner.** The specific incentives & timeouts are not considered in this document.
+
+# High level overview
+
+The `progress` function _defines_ what it means for a transition $(R, transactions) \to R'$ to be valid, in the following sense:
+
+1. From $(R, transactions)$, one can construct the "initial" machine state $M_0$.
+2. Define $M_{i+1} = progress(M_i)$
+3. Compute $M_1, M_2, ...$ until you arrive at a "terminal" state $M_t$
+4. Check that $M_t.chainState.root == R'$. If yes, $(R, transactions) \to R'$ is valid; else, it is invalid.
+
+In the event of a validator disputing a sequencer-posted state root, the "bisection game" plays out as follows.
```

Comment:
While it's less elegant, I'm not sure there's a good way to avoid 2-phase dissection.

In principle, the system could work like this:
- your initial machine state could start with an arbitrary list of transactions (they could be organized by block or not, this doesn't matter)
- your fraud proof generator would process each transaction in order, stopping when all transactions are processed
- this produces an array `[S_0, ..., S_n]` of machine states. 

This is elegant, as the dispute game can simply binary search for the first incorrect state in that list.

BUT, I would assume that producing a fraud proof for a single transaction has a non-trivial cost. The above scheme needs to know the total number of steps `n` to be safe, meaning you have to compute a fraud proof for every transaction.

With a 2-phase dispute game, where you:
1. narrow down to the first incorrectly processed transaction (if it exists)
2. construct a fraud proof for that transaction
3. narrow down to the first incorrect step in that fraud proof (if it exists)

then you only have to construct the fraud proof for a single transaction.

---

All that said, you can use the same rules for the two phases. The code here is simply solving the following problem:
- Alice has an array `a = [A_0, A_1, ..., A_n]`
- Bob has an array `b = [B_0, B_1, ..., B_n]`
- Bob forces Alice to reveal `a[i]` for certain values of `i`, comparing them to `b[i]`, allowing him to learn the first `i` so that `a[i] != b[i]`

In a 2-phase dispute game, the interactive part of both phases can be reduced to this problem.